src/frontend/src/app/core/state/proposal/proposal.service.ts

```diff
@@ -0,0 +1,61 @@
+import { Injectable } from '@angular/core';
+import { BehaviorSubject, map, switchMap } from 'rxjs';
+
+import { ListProposalsResponse } from '@cg/backend';
+import { BackendActorService } from '~core/services';
+import { isNotNil, isOk } from '~core/utils';
+import { mapOpenProposalListResponse } from './proposal.mapper';
+import { Proposal } from './proposal.model';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class ProposalService {
+  private openProposalListSubject = new BehaviorSubject<Proposal[]>([]);
+  public openProposalList$ = this.openProposalListSubject.asObservable();
+
+  private currentProposalIdSubject = new BehaviorSubject<bigint | null>(null);
+  public currentProposalId$ = this.currentProposalIdSubject.asObservable();
+
+  private lastLoadTime: Date | null = null;
+
+  public readonly currentProposal$ = this.currentProposalId$.pipe(
+    switchMap(proposalId =>
+      this.openProposalList$.pipe(
+        map(
+          proposals =>
+            proposals.find(proposal => proposal.id === proposalId) ?? null,
+        ),
+      ),
+    ),
+  );
+
+  constructor(private readonly actorService: BackendActorService) {}
+
+  public async loadOpenProposalList(): Promise<void> {
+    let getResponse: ListProposalsResponse;
+
+    if (
+      !isNotNil(this.lastLoadTime) ||
+      new Date().getTime() - this.lastLoadTime.getTime() > 5_000
```

Comment:
I think this can be moved out to a separate function for readability:
```typesript
const CACHE_TTL = 5_000;

private isCached(): boolean {
  if (isNil(this.lastLoadTime)) {
    return false;
  }

  const cacheExpiryTime = this.lastLoadTime + CACHE_TTL;
  const currentTime = Date.now();

  return currentTime > cacheExpiryTime;
}
```

Note:
- `Date.now()` is the same as `new Date().getTime()` but a little nicer to write
- `!isNotNil` is the same as `isNil`
- `isNil || conditionB` is less efficient than `isNotNil && conditionB` due to short-circuiting
- Since we're also working with the UNIX timestamp returned from `.getTime()` we should just save that directly (comment below)